### ISSUE TITLE
Data race when using aws clients

### ISSUE THREAD

**Issue by PaulLiang1** [aws/aws-xray-sdk-go#118]
Reproduced on:
- go version: 1.11.4
- goos: linux
- goarch: amd64
- aws-xray-sdk-go@v1.0.0-rc.11 
- aws-sdk-go@v1.17.12

when running the application with `-race` flag on it spits warning:
```
app_1   | ==================
app_1   | WARNING: DATA RACE
app_1   | Read at 0x00c00059fd10 by goroutine 91:
app_1   |   reflect.Value.Bool()
app_1   |       /usr/local/go/src/reflect/value.go:254 +0x51
app_1   |   encoding/json.isEmptyValue()
app_1   |       /usr/local/go/src/encoding/json/encode.go:319 +0x1d0
app_1   |   encoding/json.(*structEncoder).encode()
app_1   |       /usr/local/go/src/encoding/json/encode.go:636 +0x344
app_1   |   encoding/json.(*structEncoder).encode-fm()
app_1   |       /usr/local/go/src/encoding/json/encode.go:661 +0x7b
app_1   |   encoding/json.(*ptrEncoder).encode()
app_1   |       /usr/local/go/src/encoding/json/encode.go:788 +0x135
app_1   |   encoding/json.(*ptrEncoder).encode-fm()
app_1   |       /usr/local/go/src/encoding/json/encode.go:793 +0x7b
app_1   |   encoding/json.(*encodeState).reflectValue()
app_1   |       /usr/local/go/src/encoding/json/encode.go:333 +0x93
app_1   |   encoding/json.(*encodeState).marshal()
app_1   |       /usr/local/go/src/encoding/json/encode.go:305 +0xad
app_1   |   encoding/json.Marshal()
app_1   |       /usr/local/go/src/encoding/json/encode.go:160 +0x73
app_1   |   github.com/aws/aws-xray-sdk-go/xray.packSegments.func1()
app_1   |       /go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0-rc.11/xray/default_emitter.go:101 +0x318
app_1   |   github.com/aws/aws-xray-sdk-go/xray.packSegments()
app_1   |       /go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0-rc.11/xray/default_emitter.go:107 +0x29b
app_1   |   github.com/aws/aws-xray-sdk-go/xray.(*DefaultEmitter).Emit()
app_1   |       /go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0-rc.11/xray/default_emitter.go:64 +0xd0
app_1   |   github.com/aws/aws-xray-sdk-go/xray.(*Segment).emit()
app_1   |       /go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0-rc.11/xray/segment.go:297 +0xc5
app_1   |   github.com/aws/aws-xray-sdk-go/xray.(*Segment).flush()
app_1   |       /go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0-rc.11/xray/segment.go:314 +0x241
app_1   |   github.com/aws/aws-xray-sdk-go/xray.(*Segment).safeFlush()
app_1   |       /go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0-rc.11/xray/segment.go:330 +0xb1
app_1   |   github.com/aws/aws-xray-sdk-go/xray.(*Segment).flush()
app_1   |       /go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0-rc.11/xray/segment.go:321 +0xcf
app_1   |   github.com/aws/aws-xray-sdk-go/xray.(*Segment).Close()
app_1   |       /go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0-rc.11/xray/segment.go:250 +0x238
app_1   |   github.com/aws/aws-xray-sdk-go/xray.xrayCompleteHandler.func1()
app_1   |       /go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0-rc.11/xray/aws.go:220 +0xb5e
app_1   |   github.com/aws/aws-sdk-go/aws/request.(*HandlerList).Run()
app_1   |       /go/pkg/mod/github.com/aws/aws-sdk-go@v1.17.12/aws/request/handlers.go:216 +0xd3
app_1   |   github.com/aws/aws-sdk-go/aws/request.(*Request).Send.func1()
app_1   |       /go/pkg/mod/github.com/aws/aws-sdk-go@v1.17.12/aws/request/request.go:464 +0x46
app_1   |   github.com/aws/aws-sdk-go/aws/request.(*Request).Send()
app_1   |       /go/pkg/mod/github.com/aws/aws-sdk-go@v1.17.12/aws/request/request.go:483 +0x31d
app_1   |   github.com/aws/aws-sdk-go/service/dynamodb.(*DynamoDB).QueryWithContext()
app_1   |       /go/pkg/mod/github.com/aws/aws-sdk-go@v1.17.12/service/dynamodb/api.go:3139 +0x232
            ...

app_1   |
app_1   | Previous write at 0x00c00059fd10 by goroutine 90:
app_1   |   github.com/aws/aws-xray-sdk-go/xray.(*Segment).addError()
app_1   |       /go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0-rc.11/xray/segment.go:445 +0x5c
app_1   |   github.com/aws/aws-xray-sdk-go/xray.(*Segment).Close()
app_1   |       /go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0-rc.11/xray/segment.go:247 +0x268
app_1   |   github.com/aws/aws-xray-sdk-go/xray.xrayCompleteHandler.func1()
app_1   |       /go/pkg/mod/github.com/aws/aws-xray-sdk-go@v1.0.0-rc.11/xray/aws.go:220 +0xb5e
app_1   |   github.com/aws/aws-sdk-go/aws/request.(*HandlerList).Run()
app_1   |       /go/pkg/mod/github.com/aws/aws-sdk-go@v1.17.12/aws/request/handlers.go:216 +0xd3
app_1   |   github.com/aws/aws-sdk-go/aws/request.(*Request).Send.func1()
app_1   |       /go/pkg/mod/github.com/aws/aws-sdk-go@v1.17.12/aws/request/request.go:464 +0x46
app_1   |   github.com/aws/aws-sdk-go/aws/request.(*Request).Send()
app_1   |       /go/pkg/mod/github.com/aws/aws-sdk-go@v1.17.12/aws/request/request.go:483 +0x31d
app_1   |   github.com/aws/aws-sdk-go/service/dynamodb.(*DynamoDB).QueryWithContext()
app_1   |       /go/pkg/mod/github.com/aws/aws-sdk-go@v1.17.12/service/dynamodb/api.go:3139 +0x232
            ...
```

Removing any patching for aws sdk makes the problem go away
```
dynamoClient := dynamodb.New(awsSession)
lambdaClient := lambda.New(awsSession)

//xray.AWS(dynamoClient.Client)
//xray.AWS(lambdaClient.Client)
```

**Comment by luluzhao** [comment]
@PaulLiang1 thank you for reporting this issue. We will add it into our backlog and prioritize it. In the meantime, A PR is also welcome.

**Comment by PaulLiang1** [comment]
Any updates? 
As a fee paying customer we had to remove go xray sdk from our projects as it's producing noise for all data race warning (we split some requests to run with data race detector)

**Comment by awssandra** [comment]
Hi PaulLiang1,

Sorry for the issues and the late response. The team is currently looking into this. We've identified a few similar issues in this regard and hope to have a more definitive root cause and fix shortly. We will post here with any investigation findings in the mean time.

Thank you for your patience,
Sandra

**Comment by awssandra** [comment]
Hi PaulLiang1,

While we continue our efforts, please feel free to open a support case with AWS. We'd be happy to  help in any way we can.

Thanks,
Sandra

**Comment by PaulLiang1** [comment]
Hi @awssandra ,

A support case had been created 
https://console.aws.amazon.com/support/cases?region=us-west-2#/6265603331/en

Thanks,

**Comment by PaulLiang1** [comment]
Hi @awssandra ,

Some code to help reproduce the issue can be found here
https://github.com/PaulLiang1/xray-data-race-reproduce

**Comment by yogiraj07** [comment]
Hi @PaulLiang1 ,
Thank you for the repro. I am actively looking into this issue. Will keep you posted.

**Comment by yogiraj07** [comment]
Hi @PaulLiang1 ,
This has been fixed in release 1.0.0-rc.14 :  [changelog](https://github.com/aws/aws-xray-sdk-go/blob/master/CHANGELOG.md). Please feel free to reopen the issue, if you have any questions.

Thanks,
Yogi
